Batteries/Data/Vector/Lemmas.lean

```diff
@@ -96,3 +98,31 @@ theorem getElem_tail {v : Vector α n} {i : Nat} (hi : i < n - 1) :
 -- This is not a `@[simp]` lemma because the LHS simplifies to `Vector.extract`.
 theorem get_tail (v : Vector α (n + 1)) (i : Fin n) :
     v.tail.get i = v.get i.succ := getElem_tail _
+
+/-! ### finIdxOf? lemmas -/
+
+@[simp]
+theorem finIdxOf?_empty [BEq α] (v : Vector α 0) : v.finIdxOf? a = none := by
+  simp [v.eq_empty]
+
+@[simp]
+theorem finIdxOf?_eq_none_iff [BEq α] [LawfulBEq α] {v : Vector α n} {a : α} :
+    v.finIdxOf? a = none ↔ a ∉ v := by
+  obtain ⟨xs, rfl⟩ := v
+  simp
+
+@[simp]
+theorem finIdxOf?_eq_some_iff [BEq α] [LawfulBEq α] {v : Vector α n} {a : α} {i : Fin n} :
+    v.finIdxOf? a = some i ↔ v.get i = a ∧ ∀ j < i, ¬v.get j = a := by
+  obtain ⟨xs, rfl⟩ := v
+  simp
+
+@[simp]
+theorem isSome_finIdxOf? [BEq α] [LawfulBEq α] {v : Vector α n} {a : α} :
+    (v.finIdxOf? a).isSome ↔ a ∈ v := by
+  obtain ⟨v, rfl⟩ := v
+  simp
+
+theorem isNone_finIdxOf? [BEq α] [LawfulBEq α] {v : Vector α n} {a : α} :
+    (v.finIdxOf? a).isNone = ¬ a ∈ v := by
+  simp
```

Comment:
Should this be `↔` instead of `=`? Doesn't make a real difference but with `=` I would expect `v` to be explicit in order to allow for dot-notation to work. Perhaps we should have both?

Reply:
`isNone` is a boolean, so `=` is the stronger lemma.

Reply:
The argument implicitness is copied from the `Array` API.

Reply:
I think `isSome_finIdxOf?` and `isSome_finIdxOf?` should match. All things being equal, I prefer `Iff` since I don't understand why `((v.finIdxOf? a).isNone = true) = ¬a ∈ v` is any stronger than the `Iff` version.

Reply:
I've made https://github.com/leanprover/lean4/pull/8678 to fix this inconsistency upstream.

Reply:
I've forward-ported this here, which now means the `=` is in `Bool` not `Prop`.